### PR TITLE
chore(CI/CD): collect code coverage from PKCS#11 test

### PR DESCRIPTION
**Pull request by 0lmi**
Changelog: None
Ticket: MEN-5833
Signed-off-by: Alex Miliukov <oleksandr.miliukov@northern.tech>

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -97,10 +114,11 @@ test:smoketests:linux:
   image: golang:1.18-buster
   before_script:
     - apt-get update && apt-get install -q -y make liblzma-dev libssl-dev
-    - bash tests/test_sign_with_hsm/test_sign_with_hsm.sh --setup
+    - ./tests/test_sign_with_hsm/test_sign_with_hsm.sh --setup
+    - export TEST_MENDER_ARTIFACT_PATH=./mender-artifact-linux
```
Comment (oleorhagen) [inline .gitlab-ci.yml:118]: Question: Is this just done for the aesthetics?

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -111,6 +129,32 @@ test:smoketests:linux:
...
+  before_script:
+    - ./tests/test_sign_with_hsm/test_sign_with_hsm.sh --setup
+    - export TEST_MENDER_ARTIFACT_PATH=./mender-artifact-linux-coverage
+  script:
+    - ./tests/test_sign_with_hsm/test_sign_with_hsm.sh
+  after_script:
+    - set -- /tmp/coverage-artifact*.out
+    - head -n 1 $1 > ${COVERAGE_FILE}
+    - |
+      for cover in $@; do
```
Comment (oleorhagen) [inline .gitlab-ci.yml:149]: Q: Is it not just as easy to loop over the results from the wildcard match above here?

File: cli/read.go [hunk 3]
```diff
@@ -162,7 +162,7 @@ func printPayload(index int, p handlers.Installer) {
 	fmt.Printf("    Provides:")
 	if err != nil {
 		fmt.Printf(" Invalid provides section: %s\n", err.Error())
-	} else if provides == nil || len(provides) == 0 {
+	} else if len(provides) == 0 {
```
Comment (oleorhagen) [inline cli/read.go:165]: Q: Why is this needed? :thinking:

**Review by oleorhagen — COMMENTED** [review]
Nice job, Sasha!

I think it is coming together real fine. 

I left a few comments to be discussed

File: .gitlab-ci.yml (hunk 1 above)
Comment (0lmi) [reply .gitlab-ci.yml:118]: yes. It some kind of logical to setup everything is needed in `before_script` and then in `script` just execute the test. And shebang is present in the script, so, `bash` is not needed.

File: .gitlab-ci.yml (hunk 2 above)
Comment (0lmi) [reply .gitlab-ci.yml:149]: It's also more about the aesthetics. Generally speaking we don't need `mode` in this file, but if it's cutted from a context and someone downloads an artifact, then it will be properly formatted.

**Review by kjaskiewiczz — APPROVED** [review]
looks good

File: cli/read.go (hunk 3 above)
Comment (oleorhagen) [reply cli/read.go:165]: Then please keep it in a seperate commit @0lmi  :)

File: .gitlab-ci.yml (hunk 1 above)
Comment (oleorhagen) [reply .gitlab-ci.yml:118]: It is usually considered good commit hygiene to seperate functionality from cleanups and misc other fixes.

So if this is not strictly part of the PKCS functionality, can you split it out?

File: .gitlab-ci.yml (hunk 2 above)
Comment (oleorhagen) [reply .gitlab-ci.yml:149]: Also, what happends if there are more than `9` `coverage-artifact*.out` files?

**Review by oleorhagen — APPROVED** [review]
Nice :)

**Comment by lluiscampos** [comment]
image
:partying_face:

File: .gitlab-ci.yml (hunk 1 above)
Comment (0lmi) [reply .gitlab-ci.yml:118]: 👍

File: cli/read.go (hunk 3 above)
Comment (0lmi) [reply cli/read.go:165]: 👍

File: .gitlab-ci.yml (hunk 2 above)
Comment (0lmi) [reply .gitlab-ci.yml:149]: It just works 🙃
